### PR TITLE
Remove samalba/dockerclient

### LINKED ISSUE #1879: Proposal: Roadmap to replace samalba/dockerclient with engine-api/client
## Problem

Swarm has been using [`samalba/dockerclient`](https://github.com/samalba/dockerclient) to send requests to Engines in the cluster so far. This has been a limiting factor for supporting new features of the Engine API, since `dockerclient` is outdated and does not fully support the current Engine API. 
## Solution

[`engine-api/client`](https://github.com/docker/engine-api/) contains the officially supported Docker API, along with the relevant type structs and client. We need to be using that instead.
## Challenge

All Swarm changes currently go through `dockerclient`, all of which need to be moved over to use `engine-api`, along with new using new type structs defined in `engine-api/types` and updating unit and integration tests. This is a large change that will take time to fully accomplish and test for correctness.
## Proposal

The simplest way to start is to add a new `engine-api/client` object in the `Engine` struct in Swarm, which uses the same underlying HTTP connection that is used by `dockerclient`. This means that the two clients will co-exist in Swarm while using the same underlying connection.

Once this is done, we should start moving over individual API endpoints to use the new client. We must make sure that this does not affect or cause any issues to existing Swarm users, except for the addition of new features that `engine-api` allows us to implement.
## Roadmap

The proposal will be implemented in the following steps:
- [x] Create an `engine-api` client in Swarm, and use it to replace the `Version()` function with `ServerVersion()`. This will involve adding/updating Godeps, creating mock and nop clients, and updating tests. This part is handled in #1880.
- [x] Replace the `Info()` function to use its equivalent in `engine-api`. This is another small change, with several more changes likely needed, because of type mismatches between the `Info` structs in the two client libraries. For simplicity, we should perform type conversions (mostly between `int` and `int64` as needed) - this is a BAD idea in general, but only a temporary solution until some of the next few changes take effect. This part is handled in #1938.
- [x] Change the `ContainerConfig` struct in the `cluster` package to use `ContainerConfig` from `engine-api/types`. This should mostly take care of the messy type conversions required in the previous step. We will also need to change all functions that create/modify `ContainerConfig`. This change will span across a large part of the codebase, and move over complete container management to `engine-api`. This part is handled in #1983, #2127, #2128, #2194.
- [x] Starting in file `handlers.go`, focus on individual API endpoints to move them to the new client. The endpoints can be roughly grouped into the following categories
  - [x] Network modification operations (#1982, #2068)
  - [x] Image modification operations (#1982, #2056, #2132)
  - [x] Volume modification operations (#1982)
  - [x] Events handling (#2086)
  - [x] Drop support for `HostConfig` on `/start` (https://github.com/docker/swarm/pull/2756)
- [x] The final step will be to create the HTTP client independent of `dockerclient` (as opposed to what we do right now). This might need some additional pipelining since Swarm needs some extra features, such as timeout. (#2206, https://github.com/docker/swarm/pull/2757)
## Important considerations
1. Is there a way to ensure that users can roll back to use the old client if they want, especially if things are failing? Unfortunately, it's not as simple as abstracting the client object and then choosing which client to use based on the API endpoint, because the arguments and return types are all different between `dockerclient` and `engine-api` (Thanks to @dongluochen for bringing this up)
2. Backward compatibility will break, because `engine-api` and 'dockerclient` have different types, and this affects the JSON returned by the Swarm API. (Thanks to @amitshukla for bringing this up)

### PR DESCRIPTION
**Pull request by nishanttotla**
Fix #1879 

Finally, we can now get rid of the old docker client. One less dependency.

### CODE REVIEW

**Comment by nishanttotla** [comment]
Ping @allencloud can you take a look? It would be nice to merge this.

File: cluster/engine.go [hunk 1]
```diff
@@ -1080,7 +1072,7 @@ func (e *Engine) CreateContainer(config *ContainerConfig, name string, pullImage
 	e.CheckConnectionErr(err)
 	if err != nil {
 		// If the error is other than not found, abort immediately.
-		if (err != dockerclient.ErrImageNotFound && !engineapi.IsErrImageNotFound(err)) || !pullImage {
+		if (err != testErrImageNotFound && !engineapi.IsErrImageNotFound(err)) || !pullImage {
```
Comment (allencloud) [inline cluster/engine.go]: In Line 75, there is an annotation that testErrImageNotFound is used for testing. While if that, I am afraid here we do not need this error.
Will `ContainerCreate` return an error `testErrImageNotFound`?
@nishanttotla

Comment (nishanttotla) [inline cluster/engine.go]: @allencloud `testErrImageNotFound` is only useful for our unit tests. Check how it's used in `engine_test.go`. It's not an error that is returned by Docker.

File: cluster/engine_test.go [hunk 2]
```diff
@@ -470,11 +451,11 @@ func TestCreateContainer(t *testing.T) {
 		name,
 	).Return(
 		containertypes.ContainerCreateCreatedBody{},
-		dockerclient.ErrImageNotFound,
+		testErrImageNotFound,
```
Comment (nishanttotla) [inline cluster/engine_test.go]: @allencloud see this for example. If you have a better idea, let me know.

**Comment by allencloud** [comment]
Thanks for your explanation.
LGTM

**Pull request merged**